src/category.rs

```diff
@@ -0,0 +1,235 @@
+        let rows = try!(stmt.query(&[&slug]));
+        rows.iter().next()
+                   .chain_error(|| NotFound)
+                   .map(|row| Model::from_row(&row))
+    }
+
+    pub fn encodable(self) -> EncodableCategory {
+        let Category { id: _, crates_cnt, category, slug, created_at } = self;
+        EncodableCategory {
+            id: slug.clone(),
+            slug: slug.clone(),
+            created_at: ::encode_time(created_at),
+            crates_cnt: crates_cnt,
+            category: category,
+        }
+    }
+
+    pub fn update_crate(conn: &GenericConnection,
+                        krate: &Crate,
+                        categories: &[String]) -> CargoResult<Vec<String>> {
+        let old_categories = try!(krate.categories(conn));
+        let old_categories_ids: HashSet<_> = old_categories.iter().map(|cat| {
+            cat.id
+        }).collect();
+
+        // If a new category specified is not in the database, filter
+        // it out and don't add it. Return it to be able to warn about it.
+        let mut invalid_categories = vec![];
+        let new_categories: Vec<Category> = categories.iter().flat_map(|c| {
+            match Category::find_by_slug(conn, &c) {
+                Ok(cat) => Some(cat),
+                Err(_) => {
+                    invalid_categories.push(c.to_string());
+                    None
+                },
+            }
+        }).collect();
+
+        let new_categories_ids: HashSet<_> = new_categories.iter().map(|cat| {
+            cat.id
+        }).collect();
+
+        let to_rm: Vec<_> = old_categories_ids
+                                .difference(&new_categories_ids)
+                                .cloned()
+                                .collect();
+        let to_add: Vec<_> = new_categories_ids
+                                .difference(&old_categories_ids)
+                                .cloned()
+                                .collect();
+
+        if !to_rm.is_empty() {
+            try!(conn.execute("DELETE FROM crates_categories \
+                                WHERE category_id = ANY($1) \
+                                  AND crate_id = $2",
+                              &[&to_rm, &krate.id]));
+        }
+
+        if !to_add.is_empty() {
+            let insert: Vec<_> = to_add.into_iter().map(|id| {
+                format!("({}, {})", krate.id, id)
+            }).collect();
+            let insert = insert.join(", ");
+            try!(conn.execute(&format!("INSERT INTO crates_categories \
+                                        (crate_id, category_id) VALUES {}",
+                                       insert),
+                              &[]));
+        }
+
+        Ok(invalid_categories)
+    }
+
+    pub fn count_toplevel(conn: &GenericConnection) -> CargoResult<i64> {
+        let sql = format!("\
+            SELECT COUNT(*) \
+            FROM {} \
+            WHERE category NOT LIKE '%::%'",
```

Comment:
> Ah yeah if you have it on hand I wouldn't mind taking a peek.

Ugh, it looks like I didn't check it in anywhere, unfortunately. 

The gist of it is for updating the category names, we wouldn't be able to do [one bulk upsert](https://github.com/rust-lang/crates.io/pull/473/files#diff-13d46f677e11061fc2b3c241e80537b9R30), we'd have to select each category's parent category first (in order to be able to set the parent_id), which means making sure parents are in the table before children, etc.

And to get all crates in a particular category and any of its subcategories, instead of [this query](https://github.com/rust-lang/crates.io/pull/473/files#diff-7ab30501521c288e1d8e1e994692b1a2R525) we'd have to do [a recursive CTE](http://stackoverflow.com/questions/3307480/postgresql-recursive-with) like:

```

WITH RECURSIVE recursetree(id, parent_ids) AS (
    SELECT id, NULL::int[] || parent_id
    FROM categories 
    WHERE parent_id IS NULL
  UNION ALL
    SELECT 
    c.id, 
    rt.parent_ids || c.parent_id
    FROM categories c
    JOIN recursetree rt ON rt.id = c.parent_id
  )

SELECT * 
FROM crates
INNER JOIN crates_categories 
ON crates.id = crates_categories.crate_id 
WHERE crates_categories.category_id IN (
  SELECT id
  FROM recursetree
  WHERE parent_ids @> ARRAY[(
      SELECT id 
      FROM categories 
      WHERE slug = 'development-tools'
  )]
  UNION
  SELECT id
  FROM categories
  WHERE slug = 'development-tools'
);
```

There might be ways to make this a little better, but it makes me feel kind of icky :-/